twenty-first/src/util_types/merkle_tree.rs

```diff
@@ -251,59 +276,75 @@ impl MerkleTree {
     ///   This is the other part of the de-duplication.
     ///
     /// [verify]: MerkleTreeInclusionProof::verify
-    pub fn authentication_structure(&self, leaf_indices: &[usize]) -> Result<Vec<Digest>> {
+    pub fn authentication_structure(
+        &self,
+        leaf_indices: &[MerkleTreeLeafIndex],
+    ) -> Result<Vec<Digest>> {
         let num_leafs = self.num_leafs();
         let indices = Self::authentication_structure_node_indices(num_leafs, leaf_indices)?;
-        let auth_structure = indices.map(|idx| self.nodes[idx]).collect();
+        let auth_structure = indices.map(|idx| self.node(idx).unwrap()).collect();
         Ok(auth_structure)
     }
 
     pub fn root(&self) -> Digest {
         self.nodes[Self::ROOT_INDEX]
     }
 
-    pub fn num_leafs(&self) -> usize {
-        let node_count = self.nodes.len();
+    pub fn num_leafs(&self) -> MerkleTreeLeafIndex {
+        let node_count = MerkleTreeNodeIndex::try_from(self.nodes.len()).expect("usize to u64");
         debug_assert!(node_count.is_power_of_two());
         node_count / 2
     }
 
-    pub fn height(&self) -> usize {
+    pub fn height(&self) -> MerkleTreeHeight {
         let leaf_count = self.num_leafs();
         debug_assert!(leaf_count.is_power_of_two());
-        leaf_count.ilog2() as usize
+        MerkleTreeHeight::try_from(leaf_count.ilog2()).expect("log of num leafs should fit in u32")
     }
 
     /// All nodes of the Merkle tree.
-    pub fn nodes(&self) -> &[Digest] {
-        &self.nodes
+    pub fn nodes(&self) -> impl Iterator<Item = &Digest> {
+        self.nodes.iter()
```

Comment:
How about skipping the first element? The dummy node there seems like an implementation detail to me that we shouldn't expose.

Reply:
Consider this example:

```rust
if merkle_tree.nodes()[1] == expected_root {
    rejoice!("hooray!");
}
```

Which is obviously a complete violation of interface boundaries but it wouldn't surprise me if there were lines like this all over the code base. Bite the bullet?

Reply:
We also have a function called `num_nodes()` which should probably agree with the number of elements that the iterator can yield ...

Reply:
d04e8aa544ba0bd65dca201afb861848cbc5f03b bites the bullet thusly:
 - kill `nodes()` :dagger: 
 - kill `num_nodes()` :dagger:

Reply:
`tasm-lib` uses `.nodes()` in [one place](https://github.com/search?q=repo%3ATritonVM%2Ftasm-lib%20.nodes()&type=code), as far as I can tell. This one place doesn't seem critical to performance, meaning it can easily switch to iterating the nodes' indices and requesting the corresponding nodes one-by-one. :+1: